pkg/controller/kms/key/setup.go

```diff
@@ -0,0 +1,336 @@
+	if !ok {
+		return errors.New(errUnexpectedObject)
+	}
+
+	// special case: if key is scheduled for deletion, abort early and do not schedule for deletion again
+	if cr.Status.AtProvider.DeletionDate != nil {
+		return nil
+	}
+
+	req := &svcsdk.ScheduleKeyDeletionInput{
+		KeyId: aws.String(meta.GetExternalName(cr)),
+	}
+
+	if cr.Spec.ForProvider.PendingWindowInDays != nil {
+		req.PendingWindowInDays = aws.Int64(*cr.Spec.ForProvider.PendingWindowInDays)
+	}
+
+	_, err := d.client.ScheduleKeyDeletionWithContext(ctx, req)
+
+	return errors.Wrap(err, errDelete)
+}
+
+type observer struct {
+	client svcsdkapi.KMSAPI
+}
+
+func (o *observer) lateInitialize(in *svcapitypes.KeyParameters, obj *svcsdk.DescribeKeyOutput) error {
+	// Policy
+	if in.Policy == nil {
+		resPolicy, err := o.client.GetKeyPolicy(&svcsdk.GetKeyPolicyInput{
+			KeyId:      obj.KeyMetadata.KeyId,
+			PolicyName: aws.String("default"),
+		})
+
+		if err != nil {
+			return err
+		}
+
+		in.Policy = aws.LateInitializeStringPtr(in.Policy, resPolicy.Policy)
+	}
+
+	in.Enabled = aws.LateInitializeBoolPtr(in.Enabled, obj.KeyMetadata.Enabled)
+
+	if len(in.Tags) == 0 {
+		resTags, err := o.client.ListResourceTags(&svcsdk.ListResourceTagsInput{
+			KeyId: obj.KeyMetadata.KeyId,
+		})
+
+		if err != nil {
+			return err
+		}
+
+		if len(resTags.Tags) > 0 {
+			for _, t := range resTags.Tags {
+				in.Tags = append(in.Tags, &svcapitypes.Tag{
+					TagKey:   t.TagKey,
+					TagValue: t.TagValue,
+				})
+			}
+		}
+	}
+
+	return nil
+}
+
+func (o *observer) isUpToDate(cr *svcapitypes.Key, obj *svcsdk.DescribeKeyOutput) (bool, error) {
+	// Description
+	if obj.KeyMetadata.Description != nil &&
+		cr.Spec.ForProvider.Description != nil &&
+		aws.StringValue(obj.KeyMetadata.Description) != aws.StringValue(cr.Spec.ForProvider.Description) {
+		return false, nil
+	}
+
+	// Enabled
+	if !isUpToDateEnableDisable(cr) {
+		return false, nil
+	}
+
+	// KeyPolicy
+	resPolicy, err := o.client.GetKeyPolicy(&svcsdk.GetKeyPolicyInput{
+		KeyId:      aws.String(meta.GetExternalName(cr)),
+		PolicyName: aws.String("default"),
+	})
+	if err != nil {
+		return false, err
+	}
+	if aws.StringValue(cr.Spec.ForProvider.Policy) != aws.StringValue(resPolicy.Policy) {
+		return false, nil
+	}
+
+	// Tags
+	resTags, err := o.client.ListResourceTags(&svcsdk.ListResourceTagsInput{
+		KeyId: aws.String(meta.GetExternalName(cr)),
+	})
+	if err != nil {
+		return false, nil
+	}
+	return compareTags(cr.Spec.ForProvider.Tags, resTags.Tags), nil
+}
+
+// returns which AWS Tags exist in the resource tags and which are outdated and should be removed
+func diffTags(spec []*svcapitypes.Tag, current []*svcsdk.Tag) (addTags []*svcsdk.Tag, remove []*string) {
+	addMap := make(map[string]*string, len(spec))
+	for _, t := range spec {
+		addMap[aws.StringValue(t.TagKey)] = t.TagValue
+	}
+	removeMap := map[string]struct{}{}
+	for _, t := range current {
+		if addMap[aws.StringValue(t.TagKey)] == t.TagValue {
```

Comment:
```suggestion
		if aws.StringValue(addMap[aws.StringValue(t.TagKey)]) == aws.StringValue(t.TagValue) {
```

We should compare the `string` values instead of pointers.